### PR TITLE
Reduce time-tracking overhead

### PR DESCRIPTION
**Pull request by dweindl**
* Don't check `clock()` on every rhs evaluation, but only every 500.
* Avoid `clock()` completely, if there is no max time specified.

### CODE REVIEW

File: sbml_test_models/amici_models/piecewise_with_boolean_operations/piecewise_with_boolean_operations_xB.h [hunk 1]
```diff
@@ -0,0 +1 @@
+#define xB0 xB[0]
```
Comment (FFroehlich) [inline sbml_test_models/amici_models/piecewise_with_boolean_operations/piecewise_with_boolean_operations_xB.h:1]: 🤔

File: src/solver.cpp [hunk 2]
```diff
@@ -864,7 +864,12 @@ void Solver::startTimer() const
 
 bool Solver::timeExceeded() const
 {
-    auto cputime_exceed = static_cast<double>(std::clock() - starttime_) / CLOCKS_PER_SEC;
+    // 0 means infinte time, don't call `clock()` then (expensive)
```
Comment (FFroehlich) [inline src/solver.cpp:867]:
```suggestion
    // 0 means infinite time, don't call `clock()` then (expensive)
```

File: src/solver_cvodes.cpp [hunk 3]
```diff
@@ -1036,8 +1036,14 @@ static int fxdot(realtype t, N_Vector x, N_Vector xdot, void *user_data) {
     auto solver = dynamic_cast<CVodeSolver const *>(typed_udata->second);
     Expects(model);
 
-    if(solver->timeExceeded()) {
-        return AMICI_MAX_TIME_EXCEEDED;
+    // Avoid expensive syscalls by checking only every N rhs evaluations
+    static int eval_counter = 0;
+    int const interval = 1000;
+    if (++eval_counter / interval) {
```
Comment (FFroehlich) [inline src/solver_cvodes.cpp:1042]: What about rather extending `solver->timeExceeded()` by adding the counter as private member of solver and checking against that in every call?

**Review by FFroehlich — APPROVED** [review]

File: src/solver_cvodes.cpp (hunk 3 above)
Comment (dweindl) [reply src/solver_cvodes.cpp:1042]: Thought about it, but it wouldn't work with 

https://github.com/AMICI-dev/AMICI/blob/eba2cadcd4f9439c13cd19644351c6c1b7673c68/src/amici.cpp#L148

There we want to check on every call. Would have to add some force_check=True argument or a separate method.

Comment (FFroehlich) [reply src/solver_cvodes.cpp:1042]: > Thought about it, but it wouldn't work with
> 
> https://github.com/AMICI-dev/AMICI/blob/eba2cadcd4f9439c13cd19644351c6c1b7673c68/src/amici.cpp#L148
> 
> There we want to check on every call. Would have to add some force_check=True argument or a separate method.

I think I would prefer that.

**Pull request merged**